control/controller.go

```diff
@@ -204,32 +243,51 @@ func (c *Controller) enqueueQuery(q *Query) error {
 		return errors.New("failed to transition query to queueing state")
 	}
 
-	// TODO(jsternberg): We should have a real queue! It should only
-	// start the query when we know we have a minimum amount of memory
-	// available and at least one goroutine we can use. Since we don't
-	// implement a queue at the moment, the rest just fakes it by
-	// immediately switching to the execute state and then calling
-	// start which should start the query in a new goroutine and return
-	// the underlying results.
-	return c.executeQuery(q)
+	select {
+	case c.queryQueue <- q:
+	default:
+		return errors.New("queue length exceeded")
+	}
+
+	return nil
+}
+
+func (c *Controller) processQueryQueue() {
+	for {
+		select {
+		case <-c.done:
+			return
+		case q := <-c.queryQueue:
+			c.executeQuery(q)
+		}
+	}
 }
 
-func (c *Controller) executeQuery(q *Query) error {
+func (c *Controller) executeQuery(q *Query) {
 	ctx, ok := q.tryExec()
 	if !ok {
-		return errors.New("failed to transition query into executing state")
+		if q.State() == Canceled {
+			if q.Err() == nil {
+				q.setErr(errors.New("query was canceled"))
+			}
+		} else {
+			err := errors.New("failed to transition query into executing state")
+			q.setErr(err)
+			c.logger.Error("impossible state transition", zap.Error(err))
+		}
+		return
 	}
 
-	// TODO(jsternberg): Introduce memory restrictions.
 	q.alloc = new(memory.Allocator)
+	quota := c.memoryBytesQuotaPerQuery
+	q.alloc.Limit = &quota
```

Comment:
A one liner version:

```go
q.alloc.Limit = func(v float64) *float64 { return &v }(c.memoryBytesQuotaPerQuery)
```

Reply:
This one-liner crashes.  I don't understand why, but I'm going to change it back.

Reply:
Oops, never mind, I made a silly mistake.